generator/feature_segments_checker/feature_segments_checker.cpp

```diff
@@ -75,14 +79,47 @@ void WriteCSV(Cont const & cont, string const & fileName)
     fout << a.first << "," << a.second << endl;
 }
 
-/// \returns expected point altitude in meters according to linear model.
-double ExpectedPointAltitude(int16_t startAltitudeMeters, int16_t endAltitudeMeters,
-                             double distFromStartMeters, double featureLengthMeters)
+/// \returns y = k * x + b. It's an expected altitude in meters.
+double ExpectedPointAltitude(double k, double b, double distFromStartMeters)
 {
-  if (featureLengthMeters == 0.0)
-    return 0;
-  double const k = (endAltitudeMeters - startAltitudeMeters) / featureLengthMeters;
-  return startAltitudeMeters + k * distFromStartMeters;
+  return k * distFromStartMeters + b;
+}
+
+template <class T>
+long double SumWithDevisor(vector<T> const & vec, long double devisor)
+{
+  long double result = 0.0;
+  for (auto item : vec)
+    result += item/devisor;
+  return result;
+}
+
+/// \brief Calculates factors |k| and |b| of a line using linear least squares method.
+/// \returns false in case of error (e.g. if the line is parallel with ordinate axis)
+/// and true otherwise.
+bool LinearLeastSquaresFactors(vector<double> const & xs, TAltitudeVec const & ys, double & k, double & b)
+{
+  long double const distSum = accumulate(xs.begin(), xs.end(), 0.0);
+  uint64_t const altSum = accumulate(ys.begin(), ys.end(), 0);
+  long double const distSquareSum = accumulate(xs.begin(), xs.end(), 0.0,
+                                               [](double x, double y) { return x + y * y; });
+
+  if (distSum == 0.0 || distSquareSum == 0.0)
+    return false;
+
+  uint32_t const numPoints = xs.size();
+  long double const yDivX = SumWithDevisor(ys, distSum);
+  vector<double> xMulY(xs.size());
+  transform(xs.begin(), xs.end(), ys.begin(), xMulY.begin(), [](double x, double y){ return x*y; });
```

Comment:
I need the aux vector with dot product to be able to use SumWithDivisor then. 

> Why not to simply iterate over it and calculate dot product?

What do you offer? To do something like that:
```c++
long double dotDivXSquare = 0.0;
for (int i = 0; i < xs.size(); ++i)
{
  dotDivXSquare += xs[i] * ys[i] / xSquareSum;
}
```
?